arci/src/clients/joint_position_limiter.rs

```diff
@@ -0,0 +1,362 @@
+use std::{f64, ops::RangeInclusive};
+
+use tracing::debug;
+use urdf_rs::JointType;
+
+use crate::{
+    error::Error,
+    traits::{JointTrajectoryClient, TrajectoryPoint},
+    waits::WaitFuture,
+};
+
+pub struct JointPositionLimiter<C>
+where
+    C: JointTrajectoryClient,
+{
+    client: C,
+    limits: Vec<RangeInclusive<f64>>,
+    strategy: JointPositionLimiterStrategy,
+}
+
+impl<C> JointPositionLimiter<C>
+where
+    C: JointTrajectoryClient,
+{
+    pub fn new(client: C, limits: Vec<RangeInclusive<f64>>) -> Self {
+        Self::new_with_strategy(client, limits, Default::default())
+    }
+
+    pub fn new_with_strategy(
+        client: C,
+        limits: Vec<RangeInclusive<f64>>,
+        strategy: JointPositionLimiterStrategy,
+    ) -> Self {
+        assert!(client.joint_names().len() == limits.len());
+        Self {
+            client,
+            limits,
+            strategy,
+        }
+    }
+
+    pub fn from_urdf(client: C, joints: &[urdf_rs::Joint]) -> Result<Self, Error> {
+        Self::from_urdf_with_strategy(client, joints, Default::default())
+    }
+
+    pub fn from_urdf_with_strategy(
+        client: C,
+        joints: &[urdf_rs::Joint],
+        strategy: JointPositionLimiterStrategy,
+    ) -> Result<Self, Error> {
+        let mut limits = Vec::new();
+        for joint_name in client.joint_names() {
+            if let Some(i) = joints.iter().position(|j| j.name == *joint_name) {
+                let joint = &joints[i];
+                let limit = if JointType::Continuous == joint.joint_type {
```

Comment:
If the joint is continuous, it should not have limits.

See
https://wiki.ros.org/urdf/XML/joint

```
<limit> (required only for revolute and prismatic joint)
```

Reply:
done in 807d711